### ISSUE TITLE
How to catch degenerate credentials?

### ISSUE THREAD

**Issue by jennybc** [hadley/httr#250]
What is the recommended way to catch a valid `Token2.0` object that nonetheless holds invalid credentials?

Run this bit of the Google demo:

```
> library(httr)
> myapp <- oauth_app("google",
+ key = "16795585089.apps.googleusercontent.com",
+ secret = "hlJNgK73GjUXILBQvyvOyurl")
> google_token <-
+ oauth2.0_token(oauth_endpoints("google"), myapp,
+ scope = "https://www.googleapis.com/auth/userinfo.profile",
+ cache = TRUE)
Waiting for authentication in browser...
Press Esc/Ctrl + C to abort
Authentication complete.
```

BUT, in the browser, click "Cancel" instead of "Accept". The resulting token is technically valid in this sense:

```
> inherits(google_token, "Token2.0")
[1] TRUE
```

but there's a problem with the credentials:

```
> google_token$credentials
$error
[1] "invalid_request"

$error_description
[1] "Missing required parameter: code"
```

This token is still successfully cached in `.httr-oauth`. You can restart R, repeat the call to `oauth2.0_token(…)`, and get the same technically valid but degenerate token back.

How should I catch this broken state and start OAuth afresh? Right now I'm doing something like this: `if("error" %in% names(google_token$credentials)) {…}`. I guess I could write a wrapper that checks `inherits(…, "Token2.0")` and this name thing before considering the token to be valid. Or is there a better way to check for a valid `Token2.0` object *and* credentials that are at least plausible?

**Comment by hadley** [comment]
You should file a bug so that invalid tokens don't get cached ;)

**Comment by hadley** [comment]
Fixed in 21f6d577462c3855e6c645fc699cec20a003bd85
